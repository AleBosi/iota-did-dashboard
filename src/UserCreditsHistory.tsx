import React from "react";
import { getUserHistory } from "./creditUtils";

export default function UserCreditsHistory({ did }: { did: string }) {
  const events = getUserHistory(did).slice().reverse();

  if (!events.length) {
    return (
      <div className="text-gray-600 text-sm italic">
        Nessun movimento crediti.
      </div>
    );
  }

  return (
    <div className="mt-6">
      <h2 className="text-lg font-semibold mb-3">Storico crediti</h2>
      <ul className="space-y-2 font-mono text-sm text-gray-800">
        {events.map((e, i) => (
          <li
            key={i}
            className="bg-gray-100 rounded-md px-4 py-2 shadow-sm border border-gray-200 flex items-center justify-between"
          >
            <div>
              <span className="text-gray-500 mr-2">
                [{new Date(e.data).toLocaleString()}]
              </span>
              {e.descrizione}
              <span className="ml-2 text-xs text-gray-400">({e.delta})</span>
            </div>
            <div
              className={`text-lg ${
                e.tipo === "assegnazione" ? "text-green-600" : "text-red-500"
              }`}
              title={e.tipo === "assegnazione" ? "Credito assegnato" : "Credito speso"}
            >
              {e.tipo === "assegnazione" ? "➕" : "➖"}
            </div>
          </li>
        ))}
      </ul>
    </div>
  );
}
