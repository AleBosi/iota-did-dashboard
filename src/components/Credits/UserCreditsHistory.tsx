export interface CreditHistoryItem {
  id: string;
  date: string; // ISO
  from?: string; // chi ha assegnato
  to?: string;   // destinatario
  qty: number;   // quantit√† (+/-)
  reason?: string;
}

interface Props {
  history: CreditHistoryItem[];
  did?: string;
}

export default function UserCreditsHistory({ history, did }: Props) {
  return (
    <div>
      {did && (
        <div className="text-xs text-blue-600 mb-1" data-testid="main-did">
          DID: <span className="font-mono">{did}</span>
        </div>
      )}
      {(!history || history.length === 0) ? (
        <div className="text-gray-400">Nessun movimento crediti.</div>
      ) : (
        <>
          <h3 className="font-semibold mb-2">Storico movimenti crediti</h3>
          <ul className="divide-y">
            {history.map(item => (
              <li key={item.id} className="py-2 flex items-center gap-3">
                <span className={`font-mono w-20 ${item.qty > 0 ? "text-green-600" : "text-red-600"}`}>
                  {item.qty > 0 ? "+" : ""}
                  {item.qty}
                </span>
                <span className="text-xs text-gray-500 w-36">{new Date(item.date).toLocaleString()}</span>
                {item.from && (
                  <span className="text-xs text-gray-400">da: <span className="font-mono">{item.from}</span></span>
                )}
                {item.to && (
                  <span className="text-xs text-gray-400">a: <span className="font-mono">{item.to}</span></span>
                )}
                {item.reason && <span className="ml-2">{item.reason}</span>}
              </li>
            ))}
          </ul>
        </>
      )}
    </div>
  );
}
