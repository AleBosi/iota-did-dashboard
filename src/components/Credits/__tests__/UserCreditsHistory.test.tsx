import { render, screen } from "@testing-library/react";
import UserCreditsHistory, { CreditHistoryItem } from "../UserCreditsHistory";

const testDID = "did:iota:azienda-1";

const mockHistory: CreditHistoryItem[] = [
  {
    id: "1",
    date: "2025-08-05T12:00:00.000Z",
    from: "did:iota:admin",
    to: "did:iota:azienda-1",
    qty: 50,
    reason: "Acquisto",
  },
  {
    id: "2",
    date: "2025-08-06T16:00:00.000Z",
    from: "did:iota:azienda-1",
    to: "did:iota:op-7",
    qty: -20,
    reason: "Assegnazione",
  },
];

describe("UserCreditsHistory", () => {
  it("visualizza il DID sopra la lista movimenti", () => {
    render(<UserCreditsHistory history={mockHistory} did={testDID} />);
    // Cerca SOLO il div principale col data-testid
    expect(screen.getByTestId("main-did")).toHaveTextContent(`DID: ${testDID}`);
  });

  it("mostra la lista dei movimenti crediti", () => {
    render(<UserCreditsHistory history={mockHistory} did={testDID} />);
    expect(screen.getByText("+50")).toBeInTheDocument();
    expect(screen.getByText("-20")).toBeInTheDocument();
    expect(screen.getByText(/Acquisto/)).toBeInTheDocument();
    expect(screen.getByText(/Assegnazione/)).toBeInTheDocument();
  });

  it("mostra messaggio e DID quando la history Ã¨ vuota", () => {
    render(<UserCreditsHistory history={[]} did={testDID} />);
    expect(screen.getByText(/Nessun movimento crediti/i)).toBeInTheDocument();
    expect(screen.getByTestId("main-did")).toHaveTextContent(`DID: ${testDID}`);
  });

  it("non esplode se manca il did", () => {
    render(<UserCreditsHistory history={mockHistory} />);
    expect(screen.getByText("+50")).toBeInTheDocument();
    expect(screen.getByText("-20")).toBeInTheDocument();
    // Non deve mostrare la riga DID
    expect(screen.queryByTestId("main-did")).not.toBeInTheDocument();
  });

  it("snapshot", () => {
    const { asFragment } = render(<UserCreditsHistory history={mockHistory} did={testDID} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
