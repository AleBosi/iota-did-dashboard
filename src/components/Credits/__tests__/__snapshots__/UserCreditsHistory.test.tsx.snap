// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`UserCreditsHistory snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="text-xs text-blue-600 mb-1"
      data-testid="main-did"
    >
      DID: 
      <span
        class="font-mono"
      >
        did:iota:azienda-1
      </span>
    </div>
    <h3
      class="font-semibold mb-2"
    >
      Storico movimenti crediti
    </h3>
    <ul
      class="divide-y"
    >
      <li
        class="py-2 flex items-center gap-3"
      >
        <span
          class="font-mono w-20 text-green-600"
        >
          +50
        </span>
        <span
          class="text-xs text-gray-500 w-36"
        >
          8/5/2025, 2:00:00 PM
        </span>
        <span
          class="text-xs text-gray-400"
        >
          da: 
          <span
            class="font-mono"
          >
            did:iota:admin
          </span>
        </span>
        <span
          class="text-xs text-gray-400"
        >
          a: 
          <span
            class="font-mono"
          >
            did:iota:azienda-1
          </span>
        </span>
        <span
          class="ml-2"
        >
          Acquisto
        </span>
      </li>
      <li
        class="py-2 flex items-center gap-3"
      >
        <span
          class="font-mono w-20 text-red-600"
        >
          -20
        </span>
        <span
          class="text-xs text-gray-500 w-36"
        >
          8/6/2025, 6:00:00 PM
        </span>
        <span
          class="text-xs text-gray-400"
        >
          da: 
          <span
            class="font-mono"
          >
            did:iota:azienda-1
          </span>
        </span>
        <span
          class="text-xs text-gray-400"
        >
          a: 
          <span
            class="font-mono"
          >
            did:iota:op-7
          </span>
        </span>
        <span
          class="ml-2"
        >
          Assegnazione
        </span>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
